internal/discoverer/zookeeper.go

```diff
@@ -0,0 +1,259 @@
+		entryExists := false
+		for _, bindEntry := range zkService.BindEntries {
+			if entry == bindEntry {
+				entryExists = true
+				break
+			}
+		}
+		if !entryExists {
+			zkService.BindEntries = append(zkService.BindEntries, entry)
+		}
+	}
+
+	serviceInfo, _, err := zkService.WatchConn.Get(zkService.WatchPath)
+	if err != nil {
+		return err
+	}
+
+	node := Node{}
+	err = json.Unmarshal(serviceInfo, &node)
+	if err != nil {
+		return err
+	}
+
+	zd.sendMessage(zkService, []Node{node})
+
+	return nil
+}
+
+func (zd *ZookeeperDiscoverer) removeService(serviceName string) error {
+	zd.eventMutex.Lock()
+	defer zd.eventMutex.Unlock()
+	zkService, ok := zd.zkServices[serviceName]
+	if !ok {
+		return errors.New("Zookeeper service: " + serviceName + " undefined")
+	}
+
+	zd.sendMessage(zkService, make([]Node, 0))
+	zd.unsubscribe(zkService)
+
+	return nil
+}
+
+func (zd *ZookeeperDiscoverer) sendMessage(zkService *ZookeeperService, nodes []Node) {
+	messageService := &Service{
+		name:     zkService.Name,
+		nodes:    nodes,
+		entities: make(map[string]struct{}),
+	}
+	for _, entry := range zkService.BindEntries {
+		messageService.entities[entry] = struct{}{}
+	}
+
+	msg, err := messageService.NewNotifyMessage()
+	if err != nil {
+		log.Infof("Zookeeper send message fail, err: %s", err)
+		return
+	}
+
+	zd.msgCh <- msg
+}
+
+func NewZookeeperDiscoverer(disConfig interface{}) (Discoverer, error) {
+	config := disConfig.(*conf.Zookeeper)
+
+	discoverer := ZookeeperDiscoverer{
+		msgCh:      make(chan *comm.Message, 10),
+		zkConfig:   config,
+		zkServices: make(map[string]*ZookeeperService),
+		eventMutex: sync.Mutex{},
+	}
+
+	return &discoverer, nil
+}
+
+func (zd *ZookeeperDiscoverer) newZookeeperClient(serviceName string) (*ZookeeperService, error) {
+	conn, _, err := zk.Connect(zd.zkConfig.Hosts, time.Second*time.Duration(zd.zkConfig.Timeout))
+	if err != nil {
+		return nil, err
+	}
+
+	err = zd.initZookeeperRoot(conn)
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	watchPath := zd.zkConfig.Prefix + "/" + serviceName
+	service := &ZookeeperService{
+		Name:         serviceName,
+		BindEntries:  make([]string, 0),
+		WatchConn:    conn,
+		RootPath:     zd.zkConfig.Prefix,
+		WatchPath:    watchPath,
+		WatchContext: ctx,
+		WatchCancel:  cancel,
+		WatchMutex:   sync.Mutex{},
+	}
+
+	return service, nil
+}
+
+func (zd *ZookeeperDiscoverer) initZookeeperRoot(zkClient *zk.Conn) error {
+	ok, _, err := zkClient.Exists(zd.zkConfig.Prefix)
+	if err != nil {
+		return err
+	}
+
+	if !ok {
+		_, err = zkClient.Create(zd.zkConfig.Prefix, []byte(""), 0, zk.WorldACL(zk.PermAll))
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (zd *ZookeeperDiscoverer) subscribe(service *ZookeeperService) {
+	for {
+		_, _, event, err := service.WatchConn.GetW(service.WatchPath)
+		if err != nil {
+			log.Infof("subscribe service: %s, err: %s", service.WatchPath, err)
```

Comment:
Ditto. And we should return from this branch?

Reply:
not need to return. If the service has not been created in zookeeper when the service is registered, we will not be able to perceive the change of this service after returning. When the service is created in Zookeeper, it will be notified through `childEvent`.

Reply:
After your prompt, I think if the rootpath is not set we can return, what do you think?

Reply:
The `event` may be nil if the err occurs. We need to avoid accessing it.

Reply:
It has been split, please review it again.